build/xtask/src/sizes.rs

```diff
@@ -0,0 +1,158 @@
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::convert::TryInto;
+use std::path::{Path, PathBuf};
+
+use anyhow::bail;
+use goblin::Object;
+use indexmap::IndexMap;
+
+use crate::Config;
+
+fn armv7m_suggest(size: u64) -> u64 {
+    // Nearest power of two
+    let mut i = 1;
+    while i < size {
+        i *= 2;
+    }
+    i
+}
+fn armv8m_suggest(size: u64) -> u64 {
+    // Nearest chunk of 32
+    ((size + 31) / 32) * 32
+}
+
+pub fn run(cfg: &Path) -> anyhow::Result<()> {
+    let cfg_contents = std::fs::read(&cfg)?;
+    let toml: Config = toml::from_slice(&cfg_contents)?;
+
+    let mut dist_dir = PathBuf::from("target");
+    dist_dir.push(&toml.name);
+    dist_dir.push("dist");
+
+    let mut memories = IndexMap::new();
+    for (name, out) in &toml.outputs {
+        // This is called after `dist`, which logs this overflow gracefully
+        let end = out.address.checked_add(out.size).unwrap();
+        memories.insert(name.clone(), out.address..end);
+    }
+
+    let output_region = |vaddr: u64| {
+        memories
+            .iter()
+            .find(|(_, region)| region.contains(&vaddr.try_into().unwrap()))
+            .map(|(name, _)| name.as_str())
+    };
+
+    struct Task<'a> {
+        name: &'a str,
+        stacksize: u32,
+        requires: &'a IndexMap<String, u32>,
+    }
+    let mut tasks: Vec<Task> = toml
+        .tasks
+        .iter()
+        .map(|(name, task)| Task {
+            name,
+            stacksize: task
+                .stacksize
+                .unwrap_or_else(|| toml.stacksize.unwrap()),
+            requires: &task.requires,
+        })
+        .collect();
+    tasks.push(Task {
+        name: "kernel",
+        stacksize: toml.stacksize.unwrap(),
+        requires: &toml.kernel.requires,
+    });
+
+    let suggest = if toml.target.starts_with("thumbv7m")
+        || toml.target.starts_with("thumbv7em")
+    {
+        armv7m_suggest
+    } else if toml.target.starts_with("thumbv8m") {
+        armv8m_suggest
+    } else {
+        panic!("Unknown target {}", toml.target);
+    };
+
+    let mut suggestions = Vec::new();
+    let mut first = false; // used to print a newline between tasks
+    for task in tasks {
+        if !first {
+            first = true;
+        } else {
+            println!();
+        }
+        let mut elf_name = dist_dir.clone();
+        elf_name.push(task.name);
+        let buffer = std::fs::read(elf_name)?;
+        let elf = match Object::parse(&buffer)? {
+            Object::Elf(elf) => elf,
+            o => bail!("Invalid Object {:?}", o),
+        };
+
+        let mut memory_sizes = memories
+            .keys()
+            .map(|name| (name.as_str(), 0))
+            .collect::<IndexMap<_, _>>();
+        for phdr in &elf.program_headers {
+            if let Some(region) = output_region(phdr.p_vaddr) {
+                memory_sizes[region] += phdr.p_memsz;
```

Comment:
Unless I missed a compensating routine somewhere else, this isn't quite right. Segments loaded into Flash need to have their `filsz` attributed to flash and their `memsz` attributed to RAM. You can spot this case when the vaddr and paddr differ -- the filsz should be assigned to the region associated with the paddr, while the memsz with the vaddr.